src/protocol/boolean/random_bits.rs

```diff
@@ -0,0 +1,129 @@
+use crate::error::BoxError;
+use crate::ff::{Field, Int};
+use crate::protocol::boolean::BitOpStep;
+use crate::protocol::modulus_conversion::convert_shares::{ConvertShares, XorShares};
+use crate::protocol::{context::ProtocolContext, RecordId};
+use crate::secret_sharing::Replicated;
+use futures::future::try_join_all;
+
+/// This protocol generates a sequence of uniformly random sharing of bits in `F_p`.
+/// Adding these 3-way secret-sharing will yield the secret `r_i ∈ {0,1}`.
+pub struct RandomBits {}
+
+impl RandomBits {
+    #[allow(dead_code)]
+    pub async fn execute<F: Field>(
+        ctx: ProtocolContext<'_, Replicated<F>, F>,
+        record_id: RecordId,
+    ) -> Result<(Vec<Replicated<F>>, Replicated<F>), BoxError> {
+        // We assume the bit length we operate in would not exceed 255
+        #[allow(clippy::cast_possible_truncation)]
+        let l = F::Integer::BITS as u8;
+
+        // Generate a replicated random field values. We'll use these fields as
+        // the source of `l`-bit long uniformly random sequence of bits.
+        let r_p: Replicated<F> = ctx.prss().generate_replicated(record_id);
+
+        // Convert each bit to secret sharings of that bit in the target field
+        let futures = (0..l).map(|i| {
+            // again, we don't expect our prime field to be > 2^64
+            #[allow(clippy::cast_possible_truncation)]
+            let xor_shares = XorShares {
+                num_bits: l,
+                packed_bits_left: r_p.left().as_u128() as u64,
+                packed_bits_right: r_p.right().as_u128() as u64,
+            };
+            let c = ctx.narrow(&BitOpStep::Step(i as usize));
+            async move {
+                ConvertShares::new(xor_shares)
+                    .execute_one_bit(c, record_id, i as u8)
+                    .await
```

Comment:
I'm always a bit fuzzy on this - do you need to call await here if you're about to try_join_all?

Reply:
AFAIK, yes. `try_join_all` expects its argument to implement a trait with `type Output = Result`. Without `.await` here, `execute_one_bit` returns a `Future`.